File: tui/src/tui.ts
/**
 * Core TUI types and utilities.
 * Terminal writes to process.stdout (matches pi's ProcessTerminal).
 */

// ─── ANSI Escape Sequences ─────────────────────────────────────────────────

export const ESC = "\x1b";
export const CSI = ESC + "[";
export const CLEAR = `${CSI}2J`;
export const CLEAR_LINE = `${CSI}2K`;
export const CURSOR_HIDE = `${CSI}?25l`;
export const CURSOR_SHOW = `${CSI}?25h`;
export const RESET = `${CSI}m`;
export const BOLD = `${CSI}1m`;
export const DIM = `${CSI}2m`;

// Synchronized output — terminal buffers all writes between begin/end
// and flushes atomically, preventing flicker and tearing.
export const SYNC_BEGIN = "\x1b[?2026h";
export const SYNC_END = "\x1b[?2026l";


export function cursorPos(row: number, col: number): string {
  return `${CSI}${row};${col}H`;
}

export function setFg(c: number): string {
  return `${CSI}38;5;${c}m`;
}
export function setBg(c: number): string {
  return `${CSI}48;5;${c}m`;
}

// ─── Terminal ──────────────────────────────────────────────────────────────

export interface Terminal {
  start(onInput: (data: string) => void, onResize: () => void): void;
  stop(): void;
  drainInput(maxMs?: number, idleMs?: number): Promise<void>;
  write(data: string): void;
  get columns(): number;
  get rows(): number;
  get kittyProtocolActive(): boolean;
  moveBy(lines: number): void;
  hideCursor(): void;
  showCursor(): void;
  clearLine(): void;
  clearFromCursor(): void;
  clearScreen(): void;
  setTitle(title: string): void;
  setProgress(active: boolean): void;
}

import { StdinBuffer } from "./stdin-buffer.js";
import { setKittyProtocolActive } from "./keys.js";
import * as fs from "node:fs";
import * as path from "node:path";
import * as os from "node:os";

const TERMINAL_PROGRESS_KEEPALIVE_MS = 1000;
const TERMINAL_PROGRESS_ACTIVE_SEQUENCE = "\x1b]9;4;3\x07";
const TERMINAL_PROGRESS_CLEAR_SEQUENCE = "\x1b]9;4;0;\x07";

export class NodeTerminal implements Terminal {
  private wasRaw = false;
  private inputHandler?: (data: string) => void;
  private resizeHandler?: () => void;
  private _kittyProtocolActive = false;
  private _modifyOtherKeysActive = false;
  private stdinBuffer?: StdinBuffer;
  private stdinDataHandler?: (data: string) => void;
  private progressInterval?: ReturnType<typeof setInterval>;
  private exitHandler?: () => void;
  private stopped = false;

  get kittyProtocolActive(): boolean {
    return this._kittyProtocolActive;
  }

  start(onInput: (data: string) => void, onResize: () => void): void {
    this.inputHandler = onInput;
    this.resizeHandler = onResize;
    this.stopped = false;

    this.wasRaw = process.stdin.isRaw || false;
    if (process.stdin.setRawMode) {
      process.stdin.setRawMode(true);
    }
    process.stdin.setEncoding("utf8");
    process.stdin.resume();

    // Alternate screen buffer — isolates TUI from terminal scrollback
    process.stdout.write("\x1b[?1049h");

    // Enable bracketed paste mode
    process.stdout.write("\x1b[?2004h");

    // Failsafe: restore terminal on unexpected exit (crash, uncaught exception, etc.)
    // Must be synchronous — process.exit event doesn't support async.
    this.exitHandler = () => {
      if (this.stopped) return;
      // Show cursor (may be hidden by TUI)
      process.stdout.write("\x1b[?25h");
      // Disable bracketed paste mode
      process.stdout.write("\x1b[?2004l");
      // Disable Kitty keyboard protocol
      if (this._kittyProtocolActive) {
        process.stdout.write("\x1b[<u");
        this._kittyProtocolActive = false;
      }
      // Disable modifyOtherKeys fallback
      if (this._modifyOtherKeysActive) {
        process.stdout.write("\x1b[>4;0m");
        this._modifyOtherKeysActive = false;
      }
      // Clear progress indicator
      if (this.progressInterval) {
        clearInterval(this.progressInterval);
        this.progressInterval = undefined;
        process.stdout.write(TERMINAL_PROGRESS_CLEAR_SEQUENCE);
      }
      // Restore raw mode
      if (process.stdin.setRawMode) process.stdin.setRawMode(this.wasRaw);
      // Newline so shell prompt starts clean
      process.stdout.write("\r\n");
    };
    process.on("exit", this.exitHandler);

    // Set up resize handler
    process.stdout.on("resize", this.resizeHandler);

    // Refresh terminal dimensions after suspend/resume
    if (process.platform !== "win32") {
      process.kill(process.pid, "SIGWINCH");
    }

    // Query and enable Kitty keyboard protocol
    this.queryAndEnableKittyProtocol();
  }

  private setupStdinBuffer(): void {
    this.stdinBuffer = new StdinBuffer({ timeout: 10 });

    const kittyResponsePattern = /^\x1b\[\?(\d+)u$/;

    this.stdinBuffer.on("data", (sequence) => {
      // Check for Kitty protocol response
      if (!this._kittyProtocolActive) {
        const match = sequence.match(kittyResponsePattern);
        if (match) {
          this._kittyProtocolActive = true;
          setKittyProtocolActive(true);
          process.stdout.write("\x1b[>7u");
          return;
        }
      }

      if (this.inputHandler) {
        this.inputHandler(sequence);
      }
    });

    // Re-wrap paste content with bracketed paste markers
    this.stdinBuffer.on("paste", (content) => {
      if (this.inputHandler) {
        this.inputHandler(`\x1b[200~${content}\x1b[201~`);
      }
    });

    this.stdinDataHandler = (data: string) => {
      this.stdinBuffer!.process(data);
    };
  }

  private queryAndEnableKittyProtocol(): void {
    this.setupStdinBuffer();
    process.stdin.on("data", this.stdinDataHandler!);
    process.stdout.write("\x1b[?u");
    setTimeout(() => {
      if (!this._kittyProtocolActive && !this._modifyOtherKeysActive) {
        process.stdout.write("\x1b[>4;2m");
        this._modifyOtherKeysActive = true;
      }
    }, 150);
  }

  async drainInput(maxMs = 1000, idleMs = 50): Promise<void> {
    if (this._kittyProtocolActive) {
      process.stdout.write("\x1b[<u");
      this._kittyProtocolActive = false;
      setKittyProtocolActive(false);
    }
    if (this._modifyOtherKeysActive) {
      process.stdout.write("\x1b[>4;0m");
      this._modifyOtherKeysActive = false;
    }

    // Suppress Kitty response detection during drain: the deactivation
    // response \x1b[?u would otherwise be interpreted as an activation
    // response and re-enable the protocol.
    const prevKittyActive = this._kittyProtocolActive;
    this._kittyProtocolActive = true; // sentinel to block re-enable

    const previousHandler = this.inputHandler;
    this.inputHandler = undefined;

    let lastDataTime = Date.now();
    const onData = () => { lastDataTime = Date.now(); };

    process.stdin.on("data", onData);
    const endTime = Date.now() + maxMs;

    try {
      while (true) {
        const now = Date.now();
        const timeLeft = endTime - now;
        if (timeLeft <= 0) break;
        if (now - lastDataTime >= idleMs) break;
        await new Promise((resolve) => setTimeout(resolve, Math.min(idleMs, timeLeft)));
      }
    } finally {
      process.stdin.removeListener("data", onData);
      this._kittyProtocolActive = prevKittyActive;
      this.inputHandler = previousHandler;
    }
  }

  stop(): void {
    this.stopped = true;
    if (this.exitHandler) {
      process.removeListener("exit", this.exitHandler);
      this.exitHandler = undefined;
    }

    if (this.clearProgressInterval()) {
      process.stdout.write(TERMINAL_PROGRESS_CLEAR_SEQUENCE);
    }

    // Disable bracketed paste mode
    process.stdout.write("\x1b[?2004l");

    // Disable Kitty keyboard protocol
    if (this._kittyProtocolActive) {
      process.stdout.write("\x1b[<u");
      this._kittyProtocolActive = false;
      setKittyProtocolActive(false);
    }
    if (this._modifyOtherKeysActive) {
      process.stdout.write("\x1b[>4;0m");
      this._modifyOtherKeysActive = false;
    }

    // Clean up StdinBuffer
    if (this.stdinBuffer) {
      this.stdinBuffer.destroy();
      this.stdinBuffer = undefined;
    }

    if (this.stdinDataHandler) {
      process.stdin.removeListener("data", this.stdinDataHandler);
      this.stdinDataHandler = undefined;
    }
    this.inputHandler = undefined;
    if (this.resizeHandler) {
      process.stdout.removeListener("resize", this.resizeHandler);
      this.resizeHandler = undefined;
    }

    // Exit alternate screen buffer
    process.stdout.write("\x1b[?1049l");

    process.stdin.pause();

    if (process.stdin.setRawMode) {
      process.stdin.setRawMode(this.wasRaw);
    }
  }

  write(data: string): void {
    if (process.env.PI_TUI_WRITE_LOG === "1") {
      const logDir = path.join(os.homedir(), ".future", "tui");
      const logPath = path.join(logDir, "write.log");
      try { fs.mkdirSync(logDir, { recursive: true }); } catch {}
      try { fs.appendFileSync(logPath, data, { encoding: "utf8" }); } catch {}
    }
    process.stdout.write(data);
  }

  get columns(): number {
    return process.stdout.columns || Number(process.env.COLUMNS) || 80;
  }

  get rows(): number {
    return process.stdout.rows || Number(process.env.LINES) || 24;
  }

  moveBy(lines: number): void {
    if (lines > 0) {
      process.stdout.write(`\x1b[${lines}B`);
    } else if (lines < 0) {
      process.stdout.write(`\x1b[${-lines}A`);
    }
  }

  hideCursor(): void { process.stdout.write("\x1b[?25l"); }
  showCursor(): void { process.stdout.write("\x1b[?25h"); }

  clearLine(): void {
    process.stdout.write("\x1b[K");
  }

  clearFromCursor(): void {
    process.stdout.write("\x1b[J");
  }

  clearScreen(): void {
    process.stdout.write("\x1b[2J\x1b[H");
  }

  setTitle(title: string): void {
    process.stdout.write(`\x1b]0;${title}\x07`);
  }

  setProgress(active: boolean): void {
    if (active) {
      process.stdout.write(TERMINAL_PROGRESS_ACTIVE_SEQUENCE);
      if (!this.progressInterval) {
        this.progressInterval = setInterval(() => {
          process.stdout.write(TERMINAL_PROGRESS_ACTIVE_SEQUENCE);
        }, TERMINAL_PROGRESS_KEEPALIVE_MS);
      }
    } else {
      this.clearProgressInterval();
      process.stdout.write(TERMINAL_PROGRESS_CLEAR_SEQUENCE);
    }
  }

  private clearProgressInterval(): boolean {
    if (!this.progressInterval) return false;
    clearInterval(this.progressInterval);
    this.progressInterval = undefined;
    return true;
  }
}

// ─── Theme ─────────────────────────────────────────────────────────────────

export interface Theme {
  bg: number;
  fg: number;
  accent: number;
  border: number;
  selectedBg: number;
  selectedFg: number;
  dimFg: number;
  error: number;
  success: number;
}

export const DEFAULT_THEME: Theme = {
  bg: -1,
  fg: 252,
  accent: 39,
  border: 240,
  selectedBg: 38,
  selectedFg: 255,
  dimFg: 245,
  error: 160,
  success: 76,
};

// ─── Component Architecture ─────────────────────────────────────────────────

export interface Component {
  render(width: number): string[];
  handleInput?(data: string): void;
  wantsKeyRelease?: boolean;
  invalidate(): void;
}

export interface Focusable {
  focused: boolean;
}

export function isFocusable(component: Component | null): component is Component & Focusable {
  return component !== null && "focused" in component;
}

export const CURSOR_MARKER = "\x1b_pi:c\x07";

export type InputListenerResult = { consume?: boolean; data?: string } | undefined;
export type InputListener = (data: string) => InputListenerResult;

export class Container implements Component {
  children: Component[] = [];

  addChild(component: Component): void {
    this.children.push(component);
  }

  removeChild(component: Component): void {
    const index = this.children.indexOf(component);
    if (index !== -1) this.children.splice(index, 1);
  }

  clear(): void {
    this.children = [];
  }

  invalidate(): void {
    for (const child of this.children) child.invalidate();
  }

  render(width: number): string[] {
    const lines: string[] = [];
    for (const child of this.children) {
      for (const line of child.render(width)) lines.push(line);
    }
    return lines;
  }
}

// ─── Overlay Positioning ──────────────────────────────────────────────────

export type OverlayAnchor =
  | "center"
  | "top-left" | "top-right" | "top-center"
  | "bottom-left" | "bottom-right" | "bottom-center"
  | "left-center" | "right-center";

export interface OverlayMargin {
  top?: number;
  right?: number;
  bottom?: number;
  left?: number;
}

export type SizeValue = number | `${number}%`;

export function parseSizeValue(v: SizeValue, total: number): number {
  if (typeof v === "number") return v;
  if (v.endsWith("%")) {
    const pct = parseFloat(v);
    if (isNaN(pct)) return 0;
    return Math.floor((pct / 100) * total);
  }
  return 0;
}

export interface OverlayOptions {
  width?: SizeValue;
  minWidth?: number;
  maxHeight?: SizeValue;
  anchor?: OverlayAnchor;
  offsetX?: number;
  offsetY?: number;
  row?: SizeValue;
  col?: SizeValue;
  margin?: OverlayMargin | number;
  visible?: (termWidth: number, termHeight: number) => boolean;
  nonCapturing?: boolean;
}

export interface OverlayLayout {
  row: number;
  col: number;
  width: number;
  maxHeight: number;
}

export function resolveOverlayLayout(
  termWidth: number,
  termHeight: number,
  componentLines: number,
  options?: OverlayOptions,
): OverlayLayout {
  const opt = options ?? {};
  const marginNum = typeof opt.margin === "number" ? opt.margin : 0;
  const margin = typeof opt.margin === "object"
    ? { top: opt.margin.top ?? marginNum, right: opt.margin.right ?? marginNum,
        bottom: opt.margin.bottom ?? marginNum, left: opt.margin.left ?? marginNum }
    : { top: marginNum, right: marginNum, bottom: marginNum, left: marginNum };

  const availW = termWidth - margin.left - margin.right;
  const availH = termHeight - margin.top - margin.bottom;

  let w = opt.width !== undefined ? parseSizeValue(opt.width, termWidth) : Math.min(80, availW);
  if (opt.minWidth !== undefined) w = Math.max(w, opt.minWidth);
  w = Math.max(1, Math.min(w, availW));

  let maxH = opt.maxHeight !== undefined ? parseSizeValue(opt.maxHeight, termHeight) : termHeight;
  maxH = Math.min(maxH, availH);

  const effectiveH = Math.min(componentLines, maxH);
  const anchor = opt.anchor ?? "center";

  let row: number;
  let col: number;

  // Explicit row/col override anchor
  if (opt.row !== undefined) {
    row = parseSizeValue(opt.row, availH);
  } else {
    switch (anchor) {
      case "top-left": case "top-right": case "top-center":
        row = margin.top; break;
      case "bottom-left": case "bottom-right": case "bottom-center":
        row = margin.top + availH - effectiveH; break;
      default: // center, left-center, right-center
        row = margin.top + Math.floor((availH - effectiveH) / 2);
    }
  }

  if (opt.col !== undefined) {
    col = parseSizeValue(opt.col, availW);
  } else {
    switch (anchor) {
      case "top-left": case "bottom-left": case "left-center":
        col = margin.left; break;
      case "top-right": case "bottom-right": case "right-center":
        col = margin.left + availW - w; break;
      default: // center, top-center, bottom-center
        col = margin.left + Math.floor((availW - w) / 2);
    }
  }

  row += opt.offsetY ?? 0;
  col += opt.offsetX ?? 0;
  row = Math.max(0, Math.min(row, termHeight - 1));
  col = Math.max(0, Math.min(col, termWidth - 1));

  return { row, col, width: w, maxHeight: maxH };
}

export interface OverlayHandle {
  hide(): void;
  setHidden(hidden: boolean): void;
  isHidden(): boolean;
  focus(): void;
  unfocus(): void;
  isFocused(): boolean;
}

File: tui/rust/src/tui.rs
//! Core TUI types and utilities — 1:1 port of `tui/src/tui.ts` (the parts
//! that don't touch the Node terminal): ANSI escape constants, cursor/color
//! helpers, the `Component` architecture, and overlay positioning math.
//!
//! The Node `Terminal` interface + `NodeTerminal` implementation live in
//! `terminal.rs` (self-implemented backend). The `Theme` table is in
//! `theme.rs`. `InputListener` types are app-layer concerns (P2).

// ─── ANSI Escape Sequences ─────────────────────────────────────────────────

pub const ESC: &str = "\x1b";
pub const CSI: &str = "\x1b[";
pub const CLEAR: &str = "\x1b[2J";
pub const CLEAR_LINE: &str = "\x1b[2K";
pub const CURSOR_HIDE: &str = "\x1b[?25l";
pub const CURSOR_SHOW: &str = "\x1b[?25h";
pub const RESET: &str = "\x1b[m";
pub const BOLD: &str = "\x1b[1m";
pub const DIM: &str = "\x1b[2m";

// Synchronized output — terminal buffers all writes between begin/end
// and flushes atomically, preventing flicker and tearing.
pub const SYNC_BEGIN: &str = "\x1b[?2026h";
pub const SYNC_END: &str = "\x1b[?2026l";

pub fn cursor_pos(row: usize, col: usize) -> String {
    format!("{CSI}{row};{col}H")
}

pub fn set_fg(c: u8) -> String {
    format!("{CSI}38;5;{c}m")
}

pub fn set_bg(c: u8) -> String {
    format!("{CSI}48;5;{c}m")
}

// ─── Component Architecture ─────────────────────────────────────────────────

/// ANSI "pi:c" cursor marker injected into rendered input lines; the
/// renderer strips it before writing to the terminal (see `render`).
pub const CURSOR_MARKER: &str = "\x1b_pi:c\x07";

/// A renderable UI element. `render` returns display rows (ANSI-styled
/// strings); `handle_input` receives normalized key ids (see `keys.rs`) or
/// paste chunks; `invalidate` drops any cached layout.
pub trait Component: std::any::Any {
    fn render(&mut self, width: usize) -> Vec<String>;
    fn handle_input(&mut self, _data: &str) {}
    fn invalidate(&mut self) {}

    /// Port of the TS `wantsKeyRelease?: boolean` optional component field
    /// (default false). Only consulted by the app layer (P3).
    fn wants_key_release(&self) -> bool {
        false
    }

    /// Downcast support (mirrors TS duck-typing like `"focused" in component`).
    fn as_any(&self) -> &dyn std::any::Any;
}

/// Components that can receive keyboard focus (port of the TS `Focusable`
/// interface — `{ focused: boolean }`).
pub trait Focusable: Component {
    fn focused(&self) -> bool;
    fn set_focused(&mut self, focused: bool);
}

/// Port of `isFocusable(component)` — TS checks `"focused" in component` at
/// runtime; in Rust we downcast against the known focusable components.
pub fn is_focusable(component: &dyn Component) -> bool {
    component.as_any().is::<crate::components::input::Input>()
        || component
            .as_any()
            .is::<crate::components::scoped_models_selector::ScopedModelsSelector>()
}

/// Simple container that renders children top-to-bottom (port of `Container`).
#[derive(Default)]
pub struct Container {
    pub children: Vec<Box<dyn Component>>,
}

impl Container {
    pub fn new() -> Self {
        Self {
            children: Vec::new(),
        }
    }

    pub fn add_child(&mut self, component: Box<dyn Component>) {
        self.children.push(component);
    }

    /// TS `removeChild(component)` removes by object identity; in Rust we
    /// remove by index (the app layer keeps track of its own children).
    pub fn remove_child_at(&mut self, index: usize) -> Option<Box<dyn Component>> {
        if index < self.children.len() {
            Some(self.children.remove(index))
        } else {
            None
        }
    }

    pub fn clear(&mut self) {
        self.children.clear();
    }
}

impl Component for Container {
    fn render(&mut self, width: usize) -> Vec<String> {
        let mut lines = Vec::new();
        for child in &mut self.children {
            for line in child.render(width) {
                lines.push(line);
            }
        }
        lines
    }

    fn invalidate(&mut self) {
        for child in &mut self.children {
            child.invalidate();
        }
    }

    fn as_any(&self) -> &dyn std::any::Any {
        self
    }
}

// ─── Overlay Positioning ──────────────────────────────────────────────────

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum OverlayAnchor {
    Center,
    TopLeft,
    TopRight,
    TopCenter,
    BottomLeft,
    BottomRight,
    BottomCenter,
    LeftCenter,
    RightCenter,
}

// TS `anchor ?? "center"` — the default anchor. Manual impl (clippy's
// derivable_impls suggestion needs nightly for non-unit variants).
#[allow(clippy::derivable_impls)]
impl Default for OverlayAnchor {
    fn default() -> Self {
        Self::Center
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub struct OverlayMargin {
    pub top: Option<usize>,
    pub right: Option<usize>,
    pub bottom: Option<usize>,
    pub left: Option<usize>,
}

/// TS `OverlayMargin | number` — a plain number applies to all four sides.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum MarginValue {
    All(usize),
    Sides(OverlayMargin),
}

/// Port of `SizeValue` = `number | \`${number}%\``. Invalid strings parse to
/// 0 in TS; the enum only models the two valid forms.
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum SizeValue {
    Fixed(usize),
    Percent(f64),
}

pub fn parse_size_value(v: &SizeValue, total: i64) -> i64 {
    match v {
        SizeValue::Fixed(n) => *n as i64,
        SizeValue::Percent(pct) => ((pct / 100.0) * total as f64).floor() as i64,
    }
}

#[derive(Debug, Clone, Default)]
pub struct OverlayOptions {
    pub width: Option<SizeValue>,
    pub min_width: Option<usize>,
    pub max_height: Option<SizeValue>,
    pub anchor: Option<OverlayAnchor>,
    pub offset_x: Option<i64>,
    pub offset_y: Option<i64>,
    pub row: Option<SizeValue>,
    pub col: Option<SizeValue>,
    pub margin: Option<MarginValue>,
    /// `visible(termWidth, termHeight)` gate — app-layer (P2); kept as a doc
    /// note since a bare fn pointer can't capture state.
    pub non_capturing: bool,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct OverlayLayout {
    pub row: usize,
    pub col: usize,
    pub width: usize,
    pub max_height: usize,
}

pub fn resolve_overlay_layout(
    term_width: usize,
    term_height: usize,
    component_lines: usize,
    options: Option<&OverlayOptions>,
) -> OverlayLayout {
    let opt = options.cloned().unwrap_or_default();

    let (m_top, m_right, m_bottom, m_left) = match &opt.margin {
        None => (0usize, 0usize, 0usize, 0usize),
        Some(MarginValue::All(n)) => (*n, *n, *n, *n),
        Some(MarginValue::Sides(m)) => (
            m.top.unwrap_or(0),
            m.right.unwrap_or(0),
            m.bottom.unwrap_or(0),
            m.left.unwrap_or(0),
        ),
    };

    // JS number math — widths can go negative; clamp at the end like TS.
    let avail_w = term_width as i64 - m_left as i64 - m_right as i64;
    let avail_h = term_height as i64 - m_top as i64 - m_bottom as i64;

    let mut w = match opt.width {
        Some(v) => parse_size_value(&v, term_width as i64),
        None => 80.min(avail_w),
    };
    if let Some(min_width) = opt.min_width {
        w = w.max(min_width as i64);
    }
    w = 1.max(w.min(avail_w));

    let mut max_h = match opt.max_height {
        Some(v) => parse_size_value(&v, term_height as i64),
        None => term_height as i64,
    };
    max_h = max_h.min(avail_h);

    let effective_h = (component_lines as i64).min(max_h);
    let anchor = opt.anchor.unwrap_or(OverlayAnchor::Center);

    let row = match opt.row {
        Some(v) => parse_size_value(&v, avail_h),
        None => match anchor {
            OverlayAnchor::TopLeft | OverlayAnchor::TopRight | OverlayAnchor::TopCenter => {
                m_top as i64
            }
            OverlayAnchor::BottomLeft
            | OverlayAnchor::BottomRight
            | OverlayAnchor::BottomCenter => m_top as i64 + avail_h - effective_h,
            _ => m_top as i64 + ((avail_h - effective_h) as f64 / 2.0).floor() as i64,
        },
    };

    let col = match opt.col {
        Some(v) => parse_size_value(&v, avail_w),
        None => match anchor {
            OverlayAnchor::TopLeft | OverlayAnchor::BottomLeft | OverlayAnchor::LeftCenter => {
                m_left as i64
            }
            OverlayAnchor::TopRight | OverlayAnchor::BottomRight | OverlayAnchor::RightCenter => {
                m_left as i64 + avail_w - w
            }
            _ => m_left as i64 + ((avail_w - w) as f64 / 2.0).floor() as i64,
        },
    };

    let mut row = row + opt.offset_y.unwrap_or(0);
    let mut col = col + opt.offset_x.unwrap_or(0);
    // TS: Math.max(0, Math.min(row, termHeight - 1)) — clamp inside first.
    row = row.min(term_height as i64 - 1).max(0);
    col = col.min(term_width as i64 - 1).max(0);

    OverlayLayout {
        row: row as usize,
        col: col as usize,
        width: w.max(0) as usize,
        max_height: max_h.max(0) as usize,
    }
}

// ─── Tests ─────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn cursor_pos_formats_1_based() {
        assert_eq!(cursor_pos(3, 5), "\x1b[3;5H");
    }

    #[test]
    fn set_fg_bg_use_256_color() {
        assert_eq!(set_fg(151), "\x1b[38;5;151m");
        assert_eq!(set_bg(59), "\x1b[48;5;59m");
    }

    #[test]
    fn parse_size_value_fixed_and_percent() {
        assert_eq!(parse_size_value(&SizeValue::Fixed(10), 80), 10);
        assert_eq!(parse_size_value(&SizeValue::Percent(50.0), 80), 40);
        // floor: 33% of 100 = 33; 34% of 100 = 34
        assert_eq!(parse_size_value(&SizeValue::Percent(33.0), 100), 33);
        // Negative totals are allowed (JS number math).
        assert_eq!(parse_size_value(&SizeValue::Percent(50.0), -10), -5);
    }

    #[test]
    fn overlay_default_centers() {
        let layout = resolve_overlay_layout(100, 40, 10, None);
        // width = min(80, 100) = 80; row = floor((40-10)/2) = 15; col = floor((100-80)/2) = 10
        assert_eq!(
            layout,
            OverlayLayout {
                row: 15,
                col: 10,
                width: 80,
                max_height: 40
            }
        );
    }

    #[test]
    fn overlay_respects_margin_and_max_height() {
        let options = OverlayOptions {
            margin: Some(MarginValue::All(2)),
            max_height: Some(SizeValue::Percent(50.0)),
            ..Default::default()
        };
        let layout = resolve_overlay_layout(100, 40, 30, Some(&options));
        // avail 96x36; maxH = 20; effectiveH = min(30, 20) = 20
        // row = 2 + floor((36-20)/2) = 10; col = 2 + floor((96-80)/2) = 10
        assert_eq!(
            layout,
            OverlayLayout {
                row: 10,
                col: 10,
                width: 80,
                max_height: 20
            }
        );
    }

    #[test]
    fn overlay_top_left_anchor() {
        let options = OverlayOptions {
            anchor: Some(OverlayAnchor::TopLeft),
            width: Some(SizeValue::Fixed(50)),
            ..Default::default()
        };
        let layout = resolve_overlay_layout(100, 40, 10, Some(&options));
        assert_eq!(layout.row, 0);
        assert_eq!(layout.col, 0);
        assert_eq!(layout.width, 50);
    }

    #[test]
    fn overlay_bottom_right_anchor_and_offset() {
        let options = OverlayOptions {
            anchor: Some(OverlayAnchor::BottomRight),
            width: Some(SizeValue::Fixed(50)),
            offset_x: Some(-2),
            offset_y: Some(-1),
            ..Default::default()
        };
        let layout = resolve_overlay_layout(100, 40, 10, Some(&options));
        // col = 0 + 100 - 50 - 2 = 48; row = 0 + 40 - 10 - 1 = 29
        assert_eq!(
            layout,
            OverlayLayout {
                row: 29,
                col: 48,
                width: 50,
                max_height: 40
            }
        );
    }

    #[test]
    fn overlay_clamps_to_terminal_bounds() {
        // width overflows avail: w = max(1, min(80, -10)) = 1
        let options = OverlayOptions {
            width: Some(SizeValue::Fixed(200)),
            margin: Some(MarginValue::All(60)),
            ..Default::default()
        };
        let layout = resolve_overlay_layout(100, 40, 10, Some(&options));
        assert_eq!(layout.width, 1);
        assert!(layout.col <= 99);
    }

    #[test]
    fn container_renders_children_in_order() {
        struct FixedLine(String);
        impl Component for FixedLine {
            fn render(&mut self, _width: usize) -> Vec<String> {
                vec![self.0.clone()]
            }
            fn as_any(&self) -> &dyn std::any::Any {
                self
            }
        }

        let mut c = Container::new();
        c.add_child(Box::new(FixedLine("a".into())));
        c.add_child(Box::new(FixedLine("b".into())));
        assert_eq!(c.render(80), vec!["a", "b"]);
        c.clear();
        assert!(c.render(80).is_empty());
    }
}
